content/swift/concepts/protocols/terms/codingkey/codingkey.md

```diff
@@ -0,0 +1,160 @@
+property names in the Swift object.
+
+> **Note:** JSON and Swift have different naming conventions (JSON uses *snake_case* for its key identifiers, while Swift uses *camelCase* for its property identifiers)
+## Syntax
+This protocol will work when a Swift object conforms to either an `Decodable` or `Encodable` protocol.
+
+```pseudo
+struct myStruct: Decodable {
+  enum CodingKeys: String, CodingKey { case 
+    key_one, key_two
+  }
+
+  let propertyOne: <DataType>
+  let propertyTwo: <DataType>
+}
+```
+
+```pseudo
+struct myStruct: Encodable {
+  let propertyOne: <DataType>
+  let propertyTwo: <DataType>
+
+  enum CodingKeys: String, CodingKey { case 
+    key_one, key_two
+  }
+}
+```
+
+The above demonstrates the usage of an enumerator 'CodingKeys', its cases of which refer to the JSON's object keys. It also specifies:
+1. the raw value type of the keys (it's a String because all JSON keys are of a String type only),
+2. the `CodingKey` protocol to which the enumerator has to conform in order to map the JSON keys with the Swift object's properties
+
+## Syntax example using `Decodable`
+
+The example below will decode a JSON object array of famous musicians into a Swift model using the enumerator that conforms to the `CodingKey` protocol, and
+list each musician's names and active years using a `.forEach` function
+
+```swift
+import Foundation
+
+let musiciansJSON = """
+[
+    {
+        "name": "Michael Jackson",
+        "years_active": "1971-2009"
+    },
+    {
+        "name": "Madonna",
+        "years_active": "1983-present" 
+    },
+    {
+        "name": "Prince",
+        "years_active": "1978-2016"
+    } 
+]
+"""
+
+struct Musicians: Decodable {
+    enum CodingKeys: String, CodingKey { case
+        name,
+        yearsActive = "years_active"
+    }
+
+    let name: String
+    let yearsActive: String
+}
+
+let musiciansData = Data(musiciansJSON.utf8)
+let decoder = JSONDecoder()
+let musicians = try! decoder.decode([Musicians].self, from: musiciansData)
+
+musicians.forEach {
+    print("\($0.name) was famous during \($0.yearsActive)")
+}
+```
+
+> **Note:** *Notice how we managed to use the newly-created `yearsActive` property in the Swift model when JSON's key was written as `years_active`.*
+
+
+The output on the console with be a string-interpolated version of each of the `Musicians`' array element thanks to the `.forEach` function.
+> Note that we only used a trailing closure in the invoked functon to simplify the code.
+
+```shell
+    Michael Jackson was famous during 1971-2009
+    Madonna was famous during 1983-present
+    Prince was famous during 1978-2016
+```
+
+## Syntax example using `Encodable`
+
+Following the workflow of [encoding](https://www.codecademy.com/resources/docs/swift/protocols/encodable) the Swift model into a JSON data structure, let's create an
+example of displaying the **Exercises** struct parsed into JSON.
+
+```swift
+import Foundation
```

Comment:
update the indentations and comments should begin with a capital
// Create the CodingKyes enum to define the JSON keys